SCIM Workshop Blog Post (#1408)
oktadeveloper/okta-blog: _source/_posts/2023-07-28-scim-workshop.md

---
layout: blog_post
title: "Enterprise-Ready Workshop: SCIM"
author: semona-igama
by: advocate
communities: [security,javascript]
description: "In this workshop, you will add SCIM support to a sample application, so that user changes made in your app can sync to your customer's Identity Provider!"
tags: [scim, workshop]
tweets:
- ""
- ""
- ""
image: blog/scim-workshop/social.jpg
type: awareness
---

> Hello SaaS developers! Do you currently have a need to provision, de-provision, and/or sync users across services? If so, then SCIM is the right tool for you! What is SCIM, you ask? SCIM stands for System for Cross-domain Identity Management. And it is an open standard protocol that allows us to manage identities across systems as well as manage common user lifecycles. So in this workshop, our goal is to build a SCIM server in a SaaS app using this standard protocol and, in the end, connect it to an identity provider such as Okta. 

Reviewer: The intro needs to set up the context for why this is in the enterprise-ready workshop, how it's connected to the Todo app, and how it ties to the OpenID Connect workshop.

* In the OpenID Connect workshop, you got your customer's enterprise users to log in, and created users on the fly.
* This works for logging in, but what about the rest of the story?
* What happens when an employee leaves the company? They won't be able to log in via OIDC, but you probably want to remove them from your local database too.
* What if you wanted to create an "assign task" button in your todo app, you would only know about the users who have logged in so far, but not about the rest of the users at the company.

In this workshop, we'll cover the following:

{% include toc.md %}

In this workshop, you will learn how to implement SCIM and support provisioning from multiple organizations/tenants. Although we'll integrate with Okta as an example, remember that almost every identity provider on the web has SCIM support! 

# Understanding SCIM 

> {% img blog/scim-workshop/scim-diagram.jpg alt:"Diagram of SCIM workflow" %}{: .center-image }

Author: ```suggestion

```

## What problems does SCIM solve?

> There are many challenges with managing user identities across services:

Reviewer: This is very generic, rewrite this in terms that the SaaS app developer will understand, addressing them directly.

- SaaS developers want to know as soon as someone joins a company so they can set up an account 
- SaaS developers also want to know when someone leaves a company to deactivate their account and associated resources.

> - Manually managing users through CSVs

Reviewer: What is a CSV? Why is this relevant to a SaaS app developer?

Reviewer 2: @aaronpk I'm ok leaving this out. I just gathered from Dan that some companies managed users through CSVs.

> - Manually managing users through CSVs

Reviewer 2: ```suggestion

```

- Manually adding accounts at scale

SCIM, the System for Cross-domain Identity Management, is an open standard which addresses those problems. SCIM: 

> - Provides near-instant updates to the downstream system whenever someone joins, moves inside of, or leaves a company

Reviewer: ```suggestion
- Provides near-instant updates to the SaaS app whenever someone joins, moves inside of, or leaves a company
```

> - Allows management of user identities in cloud-based applications and services - i.e., you will have awareness of your users' up-to-date information

Reviewer: "your users" is too vague, we're talking about the users of the SaaS app developer's enterprise customers here, so make that clear

> - Allows management of user identities in cloud-based applications and services - i.e., you will have awareness of your users' up-to-date information

Reviewer 2: ```suggestion
- Allows management of user identities in cloud-based applications and services - i.e., you will have an awareness of your enterprise customers' users' up-to-date information
```

## Why do we love it?

SCIM is a very well-designed standard; therefore, it can be implemented across systems that are compliant. It simplifies user identity management, thereby reducing manual admin tasks vs. having to manage user lifecycles across multiple applications. It allows user identity management by providing insight into up-to-date user information vs. having no governance over user information and assigned resources across services. Above all, it contributes to improved security! And now, who doesn't want that in this day and age?

Now that I've provided you with some background on SCIM, let's move on to building the server!

# Implement SCIM in a SaaS App 

Before we build our SCIM server, let's have a plan of action. 

- Learn about functions that support SCIM: Create, Read, Update, Delete Functions and JSON formatted requests and responses
- Install the sample application and set up your development environment
- Build the SCIM Interface in the Todo sample app and test it with Postman

> 

Reviewer 2: @edunham can you please move the scim image -  _source/_assets/img/blog/scim-workshop/scim-diagram.jpg - here instead. ty!

> 

Author: ```suggestion

{% img blog/scim-workshop/scim-diagram.jpg alt:"Diagram of SCIM workflow" %}{: .center-image }
```

> SCIM is implemented as a RESTful API. It uses Create, Read, Update, and Delete endpoints. The endpoints required by the SCIM spec are: 

Reviewer: ```suggestion
SCIM is implemented as a RESTful API. It uses Create, Read, Update, and Delete endpoints. A minimal set of endpoints for basic user management in the SCIM spec include:
```

- GET /Users - Retrieve all users 
- GET /Users?filter=userName - GET User with the filter "userName" (filtering is optional)
- GET /User/id - Retrieve a specific User by id
- POST /User - Create a user
- PUT /User/id - Update user info by id
- DEL /User/id - Delete a user by id (*Note: Okta users soft delete via PATCH)

SCIM uses the [JSON](https://www.json.org/json-en.html) format for requesting and responding to and from the Identity Provider and SCIM server. 

Have you seen these endpoints before? Are they familiar to you? Perhaps you've built backend applications with these API endpoints. Let's build on that knowledge. 

Before we begin, please note that this implementation of SCIM is meant to be vendor agnostic, and I will be adding notes on how Okta implements this standard throughout the demo. More importantly, we'll refer to [the SCIM 2.0 spec](https://datatracker.ietf.org/doc/html/rfc7644) to build the server. 

## Set Up the Sample Application

> Follow (these setup instructions)[/blog/2023/07/27/enterprise-ready-getting-started] to install and run the Todo sample app. Run `node -v` and make sure you have Node version 18 or newer. 

Reviewer: ```suggestion
Follow [these setup instructions](/blog/2023/07/27/enterprise-ready-getting-started) to install and run the Todo sample app. Run `node -v` and make sure you have Node version 18 or newer. 
```

> We will build SCIM support atop the OIDC support added in the Enterprise-Ready Workshop on OpenID Connect. To start with a version of  [the sample app](https://github.com/oktadev/okta-enterprise-ready-workshops/) that already has OIDC support implemented, `git checkout oidc-complete-live`. 

Reviewer: ```suggestion
We will build SCIM support atop the OIDC support added in the Enterprise-Ready Workshop on OpenID Connect. To start with a version of  [the sample app](https://github.com/oktadev/okta-enterprise-ready-workshops/) that already has OIDC support implemented, `git checkout oidc-workshop-complete`. 
```

 @@unique([orgId, externalId])
}
```

Now that we have the user table with the necessary fields, we'll need to edit the `seed_script.ts` file. In this workshop, I'll show you how to create multiple separate organizations and users as needed. Provisioning multiple separate orgs accommodates your customers who may be using different identity providers with different SCIM clients. By giving each organization its own endpoint and API token, we ensure that no org can accidentally update or overwrite another org's users.

Let's update the users in `seed.ts`. We'll also need to hardcode `externalId` and `active` for both users. Note that `externalId`  is a unique identifier issued by the provisioning client and must be stored in the SCIM server. The `externalId` is also considered stable because a user's name and email address can change. Along with the `active` attribute, `externalId` is not a SCIM protocol core attribute, but we will need it when we connect with Okta. In Okta, the  [Okta SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#basic-user-schema) tell us that this unique identifier is the user's global id (GUID). Ref:

> We'll also set an `apikey` to an arbitrary string, because it will be used when we implement token auth to test each route as we build it.

Reviewer: ```suggestion
We'll also set an `apikey` to an random string, because it will be used when we implement token auth to test each route as we build it.
```

        data: {
            domain: 'portal.example',
            apikey: '131313'
        }
    });
    console.log('Created org Portal', org);
    const somnusUser = await prisma.user.create({
        data: {
            name: 'Somnus',
            email: 'somnus@portal.example',
            password: 'correct horse battery staple',
            orgId: org.id,
            externalId: '22',
            active: true
        },
    })
    console.log('Created user Somnus', somnusUser)

    const trinityUser = await prisma.user.create({
        data: {
            name: 'Trinity',
            email: 'trinity@portal.example',
            password: 'Zion',
            orgId: org.id,
            externalId: '23',
            active: true
        },
    })
    console.log('Created user Trinity', trinityUser)
}

main()
    .then(async () => {
        await prisma.$disconnect()
    })
    .catch(async (e) => {
        console.error(e)
        await prisma.$disconnect()
        process.exit(1)
    })
```
Now we are ready to seed the database. If you already added records to the database, you can run `npx prisma migrate reset` before running `npm run init-db`.

> A neat feature with Prisma is the option to view the user table locally. To do this, go to the root of this workshop folder, and run `npx prisma studio`. You should be redirected to a webpage where you can see all the tables you've created. 

Reviewer: ```suggestion
A neat feature with Prisma is the option to view the user table locally. To do this, go to the root of this workshop folder, and run `npx prisma studio`. Your browser should open to a web page where you can see all the tables you've created. 
```

Note: `/scim/v2` will append to every SCIM route. You can change this static path as needed, as there isn't a specific URL path explicitly mentioned by the SCIM spec. However, [Okta's SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#base-url) recommend using `/scim/v2` unless unusual factors in your environment necessitate a different URL. 

## Build the SCIM interface:

We'll need to build each CRUD endpoint required by the SCIM spec and format the responses in JSON. After that, we'll test our functions with Postman to see that they provide the responses required by the spec. 

We'll need to use express and Prisma in `scim.ts`, so import them:

```
import express from 'express';
import { Prisma, PrismaClient } from '@prisma/client';
```

Some SCIM endpoints in `scim.ts` will return user information from the Todo app's database. To easily retrieve this information from the SCIM endpoints, implement an `IUserSchema` interface and instantiate it as the defaultUserSchema. The information in the `IUserSchema` matches the SCIM spec, as you'll see soon.  

```
const prisma = new PrismaClient();

// To funnel users into their designated orgs

> const ORG_ID = 1;

Reviewer: Why is this hardcoded to 1? Needs an explanation of what's going on here. Is this a placeholder value that you should change in production?

interface IUserSchema {
    schemas: string[];
    userName?: string;
    id?: string;
    name?: {
        givenName: string;
        familyName: string;
    };
    emails?: { primary: boolean, value: string, type: string }[];
    displayName?: string;
    locale?: string;
    meta?: {
        resourceType: string;
    }
    externalId?: string;
    groups?: [];
    password?: string;
    active?: boolean;
    detail?: string;
    status?: number;
}

const defaultUserSchema: IUserSchema = {
    schemas: ['urn:ietf:params:scim:schemas:core:2.0:User'],
    locale: 'en-US',
    groups: [],
    meta: {
        resourceType: 'User'
    }
};
```

Finally, we are ready to get started! Let's work on our first CRUD endpoint. 

### Create User Function - POST /Users - 

> Let's refer to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.3) on creating a user. To create a user, the spec says that the client sends a POST request containing a "User" to the `/Users` endpoint. In response to the POST, the server signals a successful creation with an HTTP status code 201 (Created) and returns a representation of the user created. If the server determines that the creation of the requested user conflicts with existing users (e.g., a "User" resource with a duplicate "userName"), the server MUST return HTTP status code 409 (Conflict) with a "scimType" error code of "uniqueness," as per Section 3.12.

Reviewer: This needs an explanation of who "the client" is. One of the most confusing parts about SCIM is that the "client" and "server" are kind of reversed compared to what you'd normally think of. Better to mention explicitly that the IdP is the client and will be making requests to your SaaS app. Again remember we're talking to SaaS app developers, so don't be afraid to mention that specifically, there's no need to try to make the language vague enough to apply to everyone.

> Let's refer to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.3) on creating a user. To create a user, the spec says that the client sends a POST request containing a "User" to the `/Users` endpoint. In response to the POST, the server signals a successful creation with an HTTP status code 201 (Created) and returns a representation of the user created. If the server determines that the creation of the requested user conflicts with existing users (e.g., a "User" resource with a duplicate "userName"), the server MUST return HTTP status code 409 (Conflict) with a "scimType" error code of "uniqueness," as per Section 3.12.

Reviewer 2: ```suggestion
Let's refer to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.3) on creating a user. The spec says the IdP sends a POST request containing a "User" to the `/Users` endpoint to create a user. In response to the POST, the server signals a successful creation with an HTTP status code 201 (Created) and returns a representation of the user created. If the server determines that the creation of the requested user conflicts with existing users (e.g., a "User" resource with a duplicate "userName"), the server MUST return HTTP status code 409 (Conflict) with a "scimType" error code of "uniqueness," as per Section 3.12.
```

Our `/Users` endpoint can fulfill those requirements with the following code: 

```
// Create User Function
// POST /scim/v2/Users
// RFC Notes on Creating Users: https://www.rfc-editor.org/rfc/rfc7644#section-3.3
scimRoute.post('/Users', passport.authenticate('bearer'), async (req, res) => {
    // console.log('POST: /Users');
    // Format to SCIM standard
    const newUser: IUserSchema = req.body;
    console.log(req.body)
    const { emails, password } = newUser;
    const externalId = newUser.externalId;
    const active = newUser.active;
    const givenName = newUser.name?.givenName ?? 'NAME';
    const familyName = newUser.name?.familyName ?? 'MISSING';
    const displayName = `${givenName} ${familyName}`;
    // Pull out primary email address
    const emailPrimary = emails.find(email => email.primary === true);
    const email = emailPrimary.value;
    console.log('email: ', email);
    // Set displayName to name
    const name = displayName;
    // Check if the User exists in the database
    const duplicateUser = await prisma.user.findFirst({
        select: {
            id: true,
            email: true,
            name: true,
        },
        where: {

>             email,

Reviewer: Explain why you are checking `email` here rather than `externalId`

Author: @isemona why ARE you checking email? my explanation generator is drawing a blank -- is a duplicate defined as one with the same email?

            org: { id: ORG_ID }
        }
    });
    let userResponse: IUserSchema;
    let httpStatus = 201;

>     // If there is any records returned, then we have a duplicate

Reviewer: ```suggestion
    // If there are any records returned, then we have a duplicate
```

    if (duplicateUser) {
        // User Found... Error
        console.log('Account Exist ID: ', duplicateUser.id);
        userResponse = {
            schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
            detail: 'User already exists in the database: ' + duplicateUser.id,
            status: 409
        };
        httpStatus = 409;
        // res.status(200).send(userResponse);
    } else { // If we don't find one Create...
        // Create the User in the database
        const user = await prisma.user.create({
            data: {
                org: { connect: { id: ORG_ID } },
                name,
                email,

>                 password,

Reviewer: Take out the password field, we are not setting a local password for the user.

>                 password,

Reviewer 2: ```suggestion
      
```

        userResponse = {
            ...defaultUserSchema,
            id: `${user.id}`,
            userName: user.email,
            name: {
                givenName,
                familyName
            },
            emails: [{
                primary: true,
                value: user.email,
                type: "work"
            }],
            displayName: name,
            externalId: user.externalId,
            active: user.active
        };
    }
    res.status(httpStatus).json(userResponse);
});
```

With that code saved in `scim.ts`, we have the first endpoint we can interact with. Now is the best time to test our code to ensure we are on the right track. 

To test our code, we'll be using [Postman](https://www.postman.com/) to make requests to our SCIM server. I'll also be adding an external tool, [Morgan](https://github.com/expressjs/morgan), that will help log HTTP requests; by doing so, we'll be able to confirm whether requests are reaching our server. 

### Test with Postman

We'll need to set up a few things for Postman to authenticate and interact with our local server. 

#### 1 ) Add Bearer Token Auth to Secure the SCIM Routes

> When testing our routes, Postman will authenticate to the Todo app backend using a bearer token. To support token auth, first install the appropriate passport libraries. 

Reviewer: ```suggestion
When testing our routes, Postman (acting as the SCIM client) will authenticate to the Todo app backend using a bearer token. To support token auth, first install the appropriate passport libraries. 
```

passport.use(new BearerStrategy(
  async (apikey, done) => {
    const org = await prisma.org.findFirst({
      where: {
        apikey: apikey
      }
    });

    return done(null, org);
  }
));
```

In `scim.ts`, import Passport:

```
import passport from 'passport';
```

And add passport's bearer token auth to each SCIM route in `scim.ts`. For example,

```
scimRoute.post('/Users', passport.authenticate('bearer'), async (req, res)
```

#### 2) Support Okta's SCIM content-type headers

You'll also need to install [body-parser](https://www.npmjs.com/package/body-parser) for Express.js to accept `content-type application/scim+json`, because Okta sends this in the request headers instead of `content-type application/json`. We'll need this to read the request body from Okta. Note: This specific content-type header is required by [the SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.1)

On the command line, `npm install passport-http-bearer` to add the library to your project.

In `scim.ts`, import `body-parser`:
```
import bodyParser from 'body-parser';
```
And instruct the app to use `bodyParser` for the relevant content types: 

```

> app.use(bodyParser.json({ type: ['application/json', 'application/scim+json'] }));

Author: ```suggestion
app.use(bodyParser.json({ type: ['application/scim+json'] }));
```

> ```

Reviewer 2: @edunham can you please edit this to app.use(bodyParser.json({ type: 'application/scim+json' })); - basically no need to mention 'application/json'

    "groups": [],
    "password": "1mz050nq",
    "active": true
}
```

What response do you expect? If your code is working correctly, the Todo app backend will give a response with status 201 Created, and the following body: 

```
{
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User"
    ],
    "locale": "en-US",
    "groups": [],
    "meta": {
        "resourceType": "User"
    },
    "id": "5",
    "userName": "test.user@portal.example",
    "name": {
        "givenName": "Test",
        "familyName": "User"
    },
    "emails": [
        {
            "primary": true,
            "value": "test.user@portal.example",
            "type": "work"
        }
    ],
    "displayName": "Test User",
    "externalId": "00ujl29u0le5T6Aj10h7",
    "active": true
}
```

If you send the same POST request again, re-creating an existing user, what should happen? Try it, and you'll get a 409 Conflict response, with a body like this: 

```
{
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:Error"
    ],
    "detail": "User already exists in the database: 5",
    "status": 409
}
```

You can repeat this testing process for each SCIM route that you implement! 

> ### GET /Users

Reviewer: This needs an explanation of why Okta will be making a GET request to list users in your database.

To fulfill these requirements, the sample app's backend can handle GET requests to the `/Users` endpoint in `scim.ts` like this: 

```
// Retrieve Users
// GET /scim/v2/Users
// RFC Notes on Retrieving Users: https://www.rfc-editor.org/rfc/rfc7644#section-3.4.1
scimRoute.get('/Users', passport.authenticate('bearer'), async (req, res) => {

    // console.log('GET: /Users');
    // console.log(req.query.filter)
    // RFC Notes on Pagination: https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.4
    const DEFAULT_START_INDEX = '1';
    const DEFAULT_RECORD_LIMIT = '100';
    let startIndex = parseInt(req.query.startIndex as string ?? DEFAULT_START_INDEX);
    startIndex--; // Prisma starts its indexing at 0, while Okta send starting index at 1, this is to make sure we start at 0
    const recordLimit = parseInt(req.query.recordLimit as string ?? DEFAULT_RECORD_LIMIT);
    let filterQuery: string | null = req.query.filter as string ?? null;
    let filterParams: string[] = [];
    let email = null;

>     if (!!filterQuery) {

Reviewer: is the double `!!` really right here? is this a normal JS thing because it feels tricky and looks wrong. I'd suggest either rewriting it to be less tricky or explaining what's going on.

Author: standard typescript thing to cast it to a boolean -- I too find typescript hard to look at but I think teaching its operators might be out of scope for here

If you seeded your database with `seed.ts`, the result will look like this. 

```
{
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:ListResponse"
    ],
    "totalResults": 1,
    "startIndex": 1,
    "itemsPerPage": 1,
    "Resources": [
        {
            "schemas": [
                "urn:ietf:params:scim:schemas:core:2.0:User"
            ],
            "locale": "en-US",
            "groups": [],
            "meta": {
                "resourceType": "User"
            },
            "id": "2",
            "userName": "trinity@portal.example",
            "name": {
                "givenName": "Trinity"
            },
            "emails": [
                {
                    "primary": true,
                    "value": "trinity@portal.example",
                    "type": "work"
                }
            ],
            "displayName": "Trinity",
            "externalId": "23",
            "active": true
        }
    ]
}
```
### Get user by ID

> According to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.1), the endpoint `/Users/id` can retrieve details about one particular user. If a user with the specified ID exists, the server responds with HTTP status code 200 (OK) and includes the user's information in the body of the response. It should return a 404 if no user was found with the requested identifier.  How can `scim.ts` fulfill the spec's requirements to look up users by their IDs? 

Reviewer: ```suggestion
According to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.1), the endpoint `/Users/[id]` can retrieve details about one particular user. If a user with the specified ID exists, the server responds with HTTP status code 200 (OK) and includes the user's information in the body of the response. It should return a 404 if no user was found with the requested identifier.  How can `scim.ts` fulfill the spec's requirements to look up users by their IDs? 
```

        select: {
            id: true,
            email: true,
            name: true,
            externalId: true,
            active: true,
        },
        where: {
            id,
            org: { id: ORG_ID },
        }
    });
    let httpStatus = 200;
    let userResponse;
    // If no response from DB, return 404
    if (!!user) {
        const { id, name, email } = user;
        const [givenName, familyName] = name.split(" ")

        userResponse = {
            ...defaultUserSchema,
            id: id.toString(),
            userName: email,
            name: {
                givenName,
                familyName
            },
            emails: [{
                primary: true,
                value: email,
                type: 'work'
            }],
            displayName: name,
            externalId: user.externalId,
            active: user.active
        } satisfies IUserSchema;
    } else {
        httpStatus = 404;
        userResponse = `User ${id} not found`;
    };
    // Send Response
    res.status(httpStatus).json(userResponse);
    console.log(userResponse);
});
```
##### Test looking up a user by their ID

> With Postman, send a request with no body to look up the user with ID 1. What URL should you hit to look up that user? What type of request should you send? 

Reviewer: ```suggestion
With Postman, send a request to look up the user with ID 1. What URL should you hit to look up that user? What type of request should you send? 
```

A GET request to `http://localhost:3333/scim/v2/Users/1` will look up the appropriate user. A successful response should look like this: 

```
{
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User"
    ],
    "locale": "en-US",
    "groups": [],
    "meta": {
        "resourceType": "User"
    },
    "id": "1",
    "userName": "somnus@portal.example",
    "name": {
        "givenName": "Somnus"
    },
    "emails": [
        {
            "primary": true,
            "value": "somnus@portal.example",
            "type": "work"
        }
    ],
    "displayName": "Somnus",
    "externalId": "22",
    "active": true
}
```
### Support updating user info

> Section 3.5.1 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.5.1), tells us how to update a user's information if we know their ID. Users are updated by a PUT to the `/Users/id` endpoint. if the user exists, the server updates its records for the user and responds with HTTP status code 200 (OK), including the user's latest information in the body of the response. If no user with the specified ID is found, the server returns a 404.

Reviewer: ```suggestion
Section 3.5.1 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.5.1), tells us how to update a user's information if we know their ID. Users are updated by a PUT to the `/Users/[id]` endpoint. if the user exists, the server updates its records for the user and responds with HTTP status code 200 (OK), including the user's latest information in the body of the response. If no user with the specified ID is found, the server returns a 404.
```

    ],
    "locale": "en-US",
    "groups": [],
    "meta": {
        "resourceType": "User"
    },
    "id": "3",
    "userName": "newtest.user@portal.example",
    "name": {
        "givenName": "newTest",
        "familyName": "User"
    },
    "emails": [
        {
            "primary": true,
            "value": "newtest.user@portal.example",
            "type": "work"
        }
    ],
    "displayName": "Test User",
    "externalId": "00ujl29u0le5T6Aj10h7",
    "active": true
}
```

What response do you expect? You should get an HTTP 200 with a body like this: 

```
{
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User"
    ],
    "locale": "en-US",
    "groups": [],
    "meta": {
        "resourceType": "User"
    },
    "id": "3",
    "userName": "newtest.user@portal.example",
    "name": {
        "givenName": "newTest",
        "familyName": "User"
    },
    "emails": [
        {
            "primary": true,
            "value": "newtest.user@portal.example",
            "type": "work"
        }
    ],
    "displayName": "newTest User",
    "externalId": "00ujl29u0le5T6Aj10h7",
    "active": true
}
```
### Support deleting users

> Section 3.6 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.6), tells us that service providers may choose to permanently delete users with a 204 (No Content). If no user with the specified ID is found, the server returns a 404 (Not found).

Reviewer: Why would they choose to delete users? Provide concrete examples of when this happens so the SaaS app developer understands why they should delete their users.

    // console.log('PATCH: /Users/:userId'); 
    console.log(req.body);
  
     const id  = parseInt(req.params.userId);
     const active = req.body["Operations"][0]["value"]["active"]
     await prisma.user.update({
       data: {
        active
       },
       where: { id }
     });
  
     res.sendStatus(204);
   });   
```

Try sending a PATCH request with an empty body to `http://localhost:3333/scim/v2/Users/2`.
```
{
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:PatchOp"
    ],
    "Operations": [
        {
            "op": "replace",
            "value": {
                "active": false
            }
        }
    ]
}
```

Just like the delete request, you should expect to receive a 204 response; this is in line with [Okta's docs](https://developer.okta.com/docs/reference/scim/scim-20/#update-a-specific-user-patch)

## Connecting with an Identity provider

Now that we have a working SCIM application, let's test it with an Identity Provider! In this workshop, we're using Okta. 

We'll need to set up a few things for Okta to authenticate and interact with our local server. 

### 1) Give your app a public URL

> One way to give your app a public URL or IP would be to host it on a cloud instance with DNS that you control. But if that isn't an option, you can use Localtunnel to temporarily issue a public web address to your own computer. 

Reviewer: The goal is to have your production app expose these endpoints so that the IDP can make requests to your app's SCIM endpoints. "if that isn't an option" sounds weird, just say that for development purposes yo ucan use localtunnel to provide a public address to the app running on your own computer.

> One way to give your app a public URL or IP would be to host it on a cloud instance with DNS that you control. But if that isn't an option, you can use Localtunnel to temporarily issue a public web address to your own computer. 

Reviewer 2: ```suggestion
One way to give your app a public URL or IP would be to host it on a cloud instance with DNS that you control. For development purposes, you can use Localtunnel to provide a public address to the app running on your own computer.
```

> For this example, we'll use [Localtunnel](https://theboroer.github.io/localtunnel-www/). To run the tunnel, you'll start the api with the following command:

Reviewer: ```suggestion
For this example, we'll use [Localtunnel](https://localtunnel.me). To run the tunnel, you'll start the api with the following command:
```

```
npm run serve-api
```

In another terminal, start the tunnel using:

```
npx localtunnel --port 3333
```

> It will ask for permission to install the npm package; say yes. And when you start localtunnel, it will print a URL to your terminal, such as http:/unique-url-for-squirrels. You will need that URL when we create a SCIM app in Okta so that Okta can send requests to your Todo app.

Reviewer: ```suggestion
It will ask for permission to install the npm package; say yes. And when you start localtunnel, it will print a URL to your terminal, such as http://unique-url-for-squirrels. You will need that URL when we create a SCIM app in Okta so that Okta can send requests to your Todo app.
```

In your Todo app logs, what request do you expect to see from Okta when a user's information is updated? You should see the PUT request to your `/scim/v2/Users/4` endpoint shortly after saving Leo's new name. 

## Tool Recommendations for Development 
While troubleshooting and developing an application, it can be helpful to use tools that let you inspect your code's behavior closely, including the way it handles secrets. These tools helped me as I developed the sample code used in the workshop, and might help you as you build your own projects. 

The features that make these tools excellent for testing also make them terrible for use in production. In a production context, security is essential. For troubleshooting production services, integrate your application with your team's secure logging and secrets management infrastructure. 

### View requests to and from the app

> When [ngrok](https://ngrok.com/) is supported in my environment, it's my favorite tool for viewing requests to and from a local server. If ngrok is blocked, use Okta's [system logs](https://help.okta.com/en-us/Content/Topics/Reports/Reports_SysLog.htm) to see requests from the app to Okta. [Morgan](https://github.com/expressjs/morgan) can help you log inbound HTTP requests to the server.

Reviewer: It seems weird that previously localtunnel was used, and now this says ngrok is preferred?

Reviewer 2: @aaronpk, through recording the video, I experienced multiple and sporadic timeout issues - https://github.com/localtunnel/localtunnel/issues/612. And sadly, I was not able to use ngrok through our corporate network 😞 .

Author: ```suggestion
When [ngrok](https://ngrok.com/) is supported in my environment, it's my favorite tool for viewing requests to and from a local server. Unfortunately, some firewalls forbid ngrok but allow localtunnel, which is why localtunnel was used for this workshop's demonstrations. If ngrok is blocked, use Okta's [system logs](https://help.okta.com/en-us/Content/Topics/Reports/Reports_SysLog.htm) to see requests from the app to Okta. [Morgan](https://github.com/expressjs/morgan) can help you log inbound HTTP requests to the server.
```

[Wireshark](https://www.wireshark.org/) is a powerful tool for examining all network traffic, including HTTP requests. 

### Public URLs for local apps

> ngrok is my favorite solution for giving an app a public address, but it's not allowed by all networks. When ngrok isn't supported, [localtunnel]() is a good alternative. 

Reviewer: ```suggestion
ngrok is my favorite solution for giving an app a public address, but it's not allowed by all networks. When ngrok isn't supported, [localtunnel](https://localtunnel.me) is a good alternative. 
```

Follow [Visual Studio's docs](https://code.visualstudio.com/docs/editor/debugging) to find the Run And Debug view. I used the following `launch.json` configuration for Visual Studio when working on this app:

```
{
    "configurations": [
        {
            "name": "Debug Node App",
            "type": "node",
            "request": "attach",
            "port": 3333
        },
    ]
}
```
Replace the Todo app's `apps/api/webpack.config.js` with the following to facilitate debugging through Visual Studio: 

```
const { composePlugins, withNx } = require('@nx/webpack');
const path = require('path')

// Nx plugins for webpack.
module.exports = composePlugins(withNx(), (config) => {
    // Update the webpack config as needed here.
    // e.g. `config.plugins.push(new MyPlugin())`
    config.output.devtoolModuleFilenameTemplate = function (info) {
        const rel = path.relative(process.cwd(), info.absoluteResourcePath);
        return `webpack:///./${rel}`
    }
    return config
});
```

## Conclusion 

> Congratulations, you've added SCIM support to an OIDC application! Not only have you supported SCIM for a single Okta organization, but the implementation choices made throughout this workshop prioritize scalability so that the Todo app can integrate with as many additional identity providers as necessary! By giving every integration its own unique SCIM endpoint and API token, you prepare your application to seamlessly integrate with as many customers as you want. 

Reviewer: It's worth pointing out here that some identity providers that support SCIM may have implemented a slightly different subset of SCIM than Okta implements, so you will need to read the documentation for the specific provider when adding support for a second IDP.

> Congratulations, you've added SCIM support to an OIDC application! Not only have you supported SCIM for a single Okta organization, but the implementation choices made throughout this workshop prioritize scalability so that the Todo app can integrate with as many additional identity providers as necessary! By giving every integration its own unique SCIM endpoint and API token, you prepare your application to seamlessly integrate with as many customers as you want. 

Reviewer 2: ```suggestion
Congratulations, you've added SCIM support to an OIDC application! Not only have you supported SCIM for a single Okta organization, but the implementation choices made throughout this workshop prioritize scalability so that the Todo app can integrate with as many additional identity providers as necessary! By giving every integration its own unique SCIM endpoint and API token, you prepare your application to seamlessly integrate with as many customers as you want. Keep in mind, IdPs that support SCIM may implement SCIM differently, so you will need to reference the specific IdP's implementation docs.
```

---
layout: blog_post
title: "Enterprise-Ready Workshop: SCIM"
author: semona-igama
by: advocate
communities: [security,javascript]
description: "In this workshop, you will add SCIM support to a sample application, so that user changes made in your app can sync to your customer's Identity Provider!"
tags: [scim, workshop]
tweets:
- ""
- ""
- ""
image: blog/scim-workshop/social.jpg
type: awareness
---

> Hello SaaS developers! Do you currently have a need to provision, de-provision, and/or sync users across services? If so, then SCIM is the right tool for you! What is SCIM, you ask? SCIM stands for System for Cross-domain Identity Management. And it is an open standard protocol that allows us to manage identities across systems as well as manage common user lifecycles. So in this workshop, our goal is to build a SCIM server in a SaaS app using this standard protocol and, in the end, connect it to an identity provider such as Okta. 

Author: ```suggestion
Hello SaaS developers! You sell your software to technologically mature enterprises, and they expect it to interface seamlessly with all their other tools. In our Enterprise-Ready Workshop on OpenID Connect, you learned how to solve part of this problem, by creating user accounts in your application for your customers' employees whenever they log in. 

But creating accounts when users log in is only one of your customers' many expectations! Your app is also expected to know about users who haven't logged in yet, and remove the accounts of employees who are removed from your customer's identity provider. 

In this workshop, we'll solve those problems and more with SCIM, the System for Cross-domain Identity Management. SCIM is an open standard protocol that allows us to manage identities across systems as well as common user lifecycles. In this workshop, we will enhance a sample application by adding SCIM server functionality to it, and connect the sample application's SCIM server to an Okta client. 
```

In this workshop, we'll cover the following:

{% include toc.md %}

In this workshop, you will learn how to implement SCIM and support provisioning from multiple organizations/tenants. Although we'll integrate with Okta as an example, remember that almost every identity provider on the web has SCIM support! 

# Understanding SCIM 

## What problems does SCIM solve?

> There are many challenges with managing user identities across services:

Author: ```suggestion
SaaS customers have high expectations for the applications that they purchase. Customers, especially at enterprise scale, expect apps to seamlessly integrate with identity providers in order to maximize user productivity. ```

> - SaaS developers want to know as soon as someone joins a company so they can set up an account 

Author: ```suggestion
- Enterprise customers expect that onboarding a user in their identity provider will immediately create accounts for that user in all of SaaS apps where they grant that user access.
```

> - SaaS developers also want to know when someone leaves a company to deactivate their account and associated resources.

Author: ```suggestion
- For security as well as convenience, customers expect that deleting or deactivating a user account from their identity provider will delete that user's accounts from all their enterprise SaaS apps. 
```

## Why do we love it?

SCIM is a very well-designed standard; therefore, it can be implemented across systems that are compliant. It simplifies user identity management, thereby reducing manual admin tasks vs. having to manage user lifecycles across multiple applications. It allows user identity management by providing insight into up-to-date user information vs. having no governance over user information and assigned resources across services. Above all, it contributes to improved security! And now, who doesn't want that in this day and age?

Now that I've provided you with some background on SCIM, let's move on to building the server!

# Implement SCIM in a SaaS App 

Before we build our SCIM server, let's have a plan of action. 

- Learn about functions that support SCIM: Create, Read, Update, Delete Functions and JSON formatted requests and responses
- Install the sample application and set up your development environment
- Build the SCIM Interface in the Todo sample app and test it with Postman

{% img blog/scim-workshop/scim-diagram.jpg alt:"Diagram of SCIM workflow" %}{: .center-image }

> SCIM is implemented as a RESTful API. It uses Create, Read, Update, and Delete endpoints. The endpoints required by the SCIM spec are: 

Author: ```suggestion
SCIM is implemented as a RESTful API. It uses `Create`, `Read`, `Update`, and `Delete` endpoints. The endpoints required by the SCIM spec are: 
```

```
import { scimRoute } from './scim';
app.use('/scim/v2', scimRoute);
```

Note: `/scim/v2` will append to every SCIM route. You can change this static path as needed, as there isn't a specific URL path explicitly mentioned by the SCIM spec. However, [Okta's SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#base-url) recommend using `/scim/v2` unless unusual factors in your environment necessitate a different URL. 

## Build the SCIM interface:

We'll need to build each CRUD endpoint required by the SCIM spec and format the responses in JSON. After that, we'll test our functions with Postman to see that they provide the responses required by the spec. 

We'll need to use express and Prisma in `scim.ts`, so import them:

```
import express from 'express';
import { Prisma, PrismaClient } from '@prisma/client';
```

Some SCIM endpoints in `scim.ts` will return user information from the Todo app's database. To easily retrieve this information from the SCIM endpoints, implement an `IUserSchema` interface and instantiate it as the defaultUserSchema. The information in the `IUserSchema` matches the SCIM spec, as you'll see soon.  

> 

Author: ```suggestion

To simplify the example code, we'll support one org at first, by hardcoding the org ID as 1. When you're ready to support multiple SCIM clients, you can easily replace this constant value with a function to look up the correct org ID based on 
the context of the incoming request.```

```
import { scimRoute } from './scim';
app.use('/scim/v2', scimRoute);
```

Note: `/scim/v2` will append to every SCIM route. You can change this static path as needed, as there isn't a specific URL path explicitly mentioned by the SCIM spec. However, [Okta's SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#base-url) recommend using `/scim/v2` unless unusual factors in your environment necessitate a different URL. 

## Build the SCIM interface:

We'll need to build each CRUD endpoint required by the SCIM spec and format the responses in JSON. After that, we'll test our functions with Postman to see that they provide the responses required by the spec. 

We'll need to use express and Prisma in `scim.ts`, so import them:

```
import express from 'express';
import { Prisma, PrismaClient } from '@prisma/client';
```

Some SCIM endpoints in `scim.ts` will return user information from the Todo app's database. To easily retrieve this information from the SCIM endpoints, implement an `IUserSchema` interface and instantiate it as the defaultUserSchema. The information in the `IUserSchema` matches the SCIM spec, as you'll see soon.  

> To simplify the example code, we'll support one org at first, by hardcoding the org ID as 1. When you're ready to support multiple SCIM clients, you can easily replace this constant value with a function to look up the correct org ID based on 

Reviewer 2: ```suggestion
To simplify the example code, we'll support one org at first, by hardcoding the org ID as 1. When you're ready to support multiple SCIM clients, you can easily replace this constant value with a function to look up the correct org ID based on. In addition, when we receive a request to create a user from an IdP, their org's attribute value will be hardcoded to 1 when they are created in the database. Therefore, letting us know that they belong to this specific org. 
```

> To simplify the example code, we'll support one org at first, by hardcoding the org ID as 1. When you're ready to support multiple SCIM clients, you can easily replace this constant value with a function to look up the correct org ID based on. In addition, when we receive a request to create a user from an IdP, their org's attribute value will be hardcoded to 1 when they are created in the database. Therefore, letting us know that they belong to this specific org. 

Author: ```suggestion
To simplify the example code, we'll support one org at first, by hardcoding the org ID as 1. When you're ready to support multiple SCIM clients, you can easily replace this constant value with a function to look up the correct org ID based on based on the context of the request. In addition, when an IdP instructs our SCIM server to create a user, the user's org ID will be hardcoded to 1 when they are created in the database. An update to support multiple organizations would also need to update the user creation code to associate them with the correct organization, as well. 
```

Our `/Users` endpoint can fulfill those requirements with the following code: 

```
// Create User Function
// POST /scim/v2/Users
// RFC Notes on Creating Users: https://www.rfc-editor.org/rfc/rfc7644#section-3.3
scimRoute.post('/Users', passport.authenticate('bearer'), async (req, res) => {
    // console.log('POST: /Users');
    // Format to SCIM standard
    const newUser: IUserSchema = req.body;
    console.log(req.body)
    const { emails, password } = newUser;
    const externalId = newUser.externalId;
    const active = newUser.active;
    const givenName = newUser.name?.givenName ?? 'NAME';
    const familyName = newUser.name?.familyName ?? 'MISSING';
    const displayName = `${givenName} ${familyName}`;
    // Pull out primary email address
    const emailPrimary = emails.find(email => email.primary === true);
    const email = emailPrimary.value;
    console.log('email: ', email);
    // Set displayName to name
    const name = displayName;
    // Check if the User exists in the database
    const duplicateUser = await prisma.user.findFirst({
        select: {
            id: true,
            email: true,
            name: true,
        },
        where: {

>             email,

Reviewer 2: ```suggestion
            id,
```

    "externalId": "00ujl29u0le5T6Aj10h7",
    "active": true
}
```

If you send the same POST request again, re-creating an existing user, what should happen? Try it, and you'll get a 409 Conflict response, with a body like this: 

```
{
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:Error"
    ],
    "detail": "User already exists in the database: 5",
    "status": 409
}
```

You can repeat this testing process for each SCIM route that you implement! 

### GET /Users
According to [the SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.4), you may return users paginated. [Okta requests users paginated]( https://developer.okta.com/docs/reference/scim/scim-20/#retrieve-users) with a Start Index set to 1 and a count set to 100 per page. 

In addition, filtering by username is an OPTIONAL parameter for SCIM service providers refer to this section of [the SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2). However, Okta requires username filtering support as it checks to see if the user exists by username first before creating them. Ref: [Okta Docs](https://developer.okta.com/docs/reference/scim/scim-20/#create-users)

> 

Author: ```suggestion

Why would a SCIM client, in this case the identity provider, ever want to ask our app's SCIM server what users it knows about? Okta keeps track of what identifiers each SCIM integration is using for each user, and stores those as the user's "externalId" Okta uses this endpoint to test whether a user exists and find out what ID the SCIM server uses for that individual, so that it can send the correct request type with the correct ID: PUT to update an existing user, or POST to create a new account for a user that wasn't previously known to the SCIM server. 

```

    ],
    "locale": "en-US",
    "groups": [],
    "meta": {
        "resourceType": "User"
    },
    "id": "3",
    "userName": "newtest.user@portal.example",
    "name": {
        "givenName": "newTest",
        "familyName": "User"
    },
    "emails": [
        {
            "primary": true,
            "value": "newtest.user@portal.example",
            "type": "work"
        }
    ],
    "displayName": "Test User",
    "externalId": "00ujl29u0le5T6Aj10h7",
    "active": true
}
```

What response do you expect? You should get an HTTP 200 with a body like this: 

```
{
    "schemas": [
        "urn:ietf:params:scim:schemas:core:2.0:User"
    ],
    "locale": "en-US",
    "groups": [],
    "meta": {
        "resourceType": "User"
    },
    "id": "3",
    "userName": "newtest.user@portal.example",
    "name": {
        "givenName": "newTest",
        "familyName": "User"
    },
    "emails": [
        {
            "primary": true,
            "value": "newtest.user@portal.example",
            "type": "work"
        }
    ],
    "displayName": "newTest User",
    "externalId": "00ujl29u0le5T6Aj10h7",
    "active": true
}
```
### Support deleting users

> Section 3.6 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.6), tells us that service providers may choose to permanently delete users with a 204 (No Content). If no user with the specified ID is found, the server returns a 404 (Not found).

Author: ```suggestion
Section 3.6 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.6), tells us that service providers may choose to permanently delete users with a 204 (No Content). If no user with the specified ID is found, the server returns a 404 (Not found).

When a user leaves an organization, some identity providers will delete that user entirely. Revoking the user's access to all applications is an essential component of securely offboarding them. Additionally, when an enterprise purchases a subscription plan for a SaaS app with a limited number of seats, they expect that removing a user from that app will free up a seat in their license for another user to gain access.

Deleting accounts of removed users can help your application comply with data retention regulations. Over time, storing and backing up information every user who has ever used your app, instead of only tracking active accounts, can get expensive in both storage costs and backup/restore process duration. 

```
